File: jar/src/main/java/jline/api/measures/Ms_topsoe.java
package jline.api.measures;

import jline.util.matrix.Matrix;

public final class Ms_topsoe {
    private Ms_topsoe() {}

    /**
     * Topsoe distance between two probability distributions.
     * Part of Shannon's entropy family.
     *
     * @param P first probability distribution
     * @param Q second probability distribution
     * @return Topsoe distance
     */
    public static double ms_topsoe(Matrix P, Matrix Q) {
        if (P.length() != Q.length()) {
            throw new IllegalArgumentException("Distributions must have the same size");
        }

        double sum = 0.0;
        for (int i = 0; i < P.length(); i++) {
            double pi = P.get(i);
            double qi = Q.get(i);
            double denom = pi + qi;

            if (pi > 0 && denom > 0) {
                sum += pi * Math.log(2.0 * pi / denom);
            }

            if (qi > 0 && denom > 0) {
                sum += qi * Math.log(2.0 * qi / denom);
            }
        }

        return sum;
    }
}

File: jar/src/main/kotlin/jline/api/measures/Ms_topsoe.kt
package jline.api.measures

import jline.util.matrix.Matrix
import kotlin.math.ln

/**
 * Topsoe distance between two probability distributions.
 * Part of Shannon's entropy family.
 * 
 * @param P first probability distribution
 * @param Q second probability distribution
 * @return Topsoe distance
 */
fun ms_topsoe(P: Matrix, Q: Matrix): Double {
    require(P.length() == Q.length()) { "Distributions must have the same size" }
    
    var sum = 0.0
    for (i in 0 until P.length()) {
        val pi = P.get(i)
        val qi = Q.get(i)
        val denom = pi + qi
        
        if (pi > 0 && denom > 0) {
            sum += pi * ln(2.0 * pi / denom)
        }
        
        if (qi > 0 && denom > 0) {
            sum += qi * ln(2.0 * qi / denom)
        }
    }
    
    return sum
}
/**
 * Topsoe metric algorithms
 */
@Suppress("unused")
class MsTopsoeAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}